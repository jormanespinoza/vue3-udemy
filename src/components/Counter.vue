<template>
  <h2>{{ customTitle }}</h2>
  <p>
    <span data-counter="counter">{{ counter }}</span>
    <sup>{{ power }}</sup> = {{ squareCounter }}
  </p>

  <div>
    <button @click="decreaseCounter">-1</button>
    <button @click="increaseCounter">+1</button>
    <button @click="decreasePower">-1</button>
    <button @click="increasePower">+1</button>
  </div>
</template>

<script>
export default {
  name: 'Counter',
  props: {
    title: String,
    start: {
      type: Number,
      default: 25,
      validator(value) {
        return value >= 0
      }
    }
  },
  data() {
    return {
      counter: this.start,
      power: 2
    }
  },
  methods: {
    increaseCounter() {
      this.counter++
    },
    decreaseCounter() {
      this.counter--
    },
    increasePower() {
      this.power++
    },
    decreasePower() {
      this.power--
    }
  },
  computed: {
    squareCounter() {
      return this.counter ** this.power
    },
    customTitle() {
      return this.title || 'Counter'
    }
  }
}
</script>

<style>
button {
  background-color: #64bb87;
  border-radius: 5px;
  border: 1px solid white;
  color: white;
  cursor: pointer;
  margin: 0 5px;
  padding: 5px 15px;
  transition: 0.3s ease-in-out;
}

button:hover {
  background-color: #5aa67b;
  transition: 0.3s ease-in-out;
}
</style>
