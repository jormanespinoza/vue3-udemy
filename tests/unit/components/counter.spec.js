import { shallowMount, mount } from '@vue/test-utils'
import Counter from '@/components/Counter'

describe('Counter.vue', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(Counter)
  })

  test('should match snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('h2 should match default value "Counter"', () => {
    expect(wrapper.find('h2').exists()).toBeTruthy()

    const h2Value = wrapper.find('h2').text()

    expect(h2Value).toBe("Counter")
  })

  test('p tag value should match default value "25"', () => {
    const spanValue = wrapper.find('[data-counter="counter"]').text()

    expect(spanValue).toBe('25')
  })

  test('should increment counter in 1', async () => {
    const [, increaseCounterBtn] = wrapper.findAll('button')

    await increaseCounterBtn.trigger('click')

    const spanValue = wrapper.find('[data-counter="counter"]').text()

    expect(spanValue).toBe('26')
  })

  test('should decrement counter in 1', async () => {
    const [decreaseCounterBtn] = wrapper.findAll('button')

    await decreaseCounterBtn.trigger('click')

    const spanValue = wrapper.find('[data-counter="counter"]').text()

    expect(spanValue).toBe('24')
  })

  test('sup tag value should match default value "2"', () => {
    const powerValue = wrapper.find('sup').text()

    expect(powerValue).toBe('2')
  })

  test('should increment power in 1', async () => {
    const [, , , increasePowerBtn] = wrapper.findAll('button')

    await increasePowerBtn.trigger('click')

    const powerValue = wrapper.find('sup').text()

    expect(powerValue).toBe('3')
  })

  test('should decrement power in 1', async () => {
    const [, , decreasePowerBtn] = wrapper.findAll('button')

    await decreasePowerBtn.trigger('click')

    const powerValue = wrapper.find('sup').text()

    expect(powerValue).toBe('1')
  })

  test('should stablish counter default start value', () => {
    const { start } = wrapper.props()

    const counterValue = wrapper.find('[data-counter="counter"]').text()

    expect(Number(counterValue)).toBe(start)
  })
})