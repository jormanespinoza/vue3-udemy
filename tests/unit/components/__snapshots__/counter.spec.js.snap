// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter.vue should match snapshot 1`] = `
<h2>Counter</h2>
<p><span data-counter="counter">25</span><sup>2</sup> = 625</p>
<div><button>-1</button><button>+1</button><button>-1</button><button>+1</button></div>
`;
